src/lib/provable/test/twisted-curve.unit-test.ts

```diff
@@ -0,0 +1,87 @@
+import { TwistedCurveParams } from '../../../bindings/crypto/elliptic-curve-examples.js';
+import { createCurveTwisted } from '../../../bindings/crypto/elliptic-curve.js';
+import {
+  array,
+  equivalentProvable,
+  map,
+  onlyIf,
+  spec,
+  unit,
+} from '../../testing/equivalent.js';
+import { Random } from '../../testing/random.js';
+import { assert } from '../gadgets/common.js';
+import {
+  Point,
+  CurveTwisted,
+  simpleMapToCurve,
+} from '../gadgets/twisted-curve.js';
+import { foreignField, throwError } from './test-utils.js';
+
+// provable equivalence tests
+const Edwards25519 = createCurveTwisted(TwistedCurveParams.Edwards25519);
+let curves = [Edwards25519];
+
+for (let Curve of curves) {
+  // prepare test inputs
+  let field = foreignField(Curve.Field);
+  let scalar = foreignField(Curve.Scalar);
+
+  // valid random point
+  let point = spec({
+    rng: Random.map(field.rng, (x) => simpleMapToCurve(x, Curve)),
+    there: Point.from,
+    back: Point.toBigint,
+    provable: Point.provable,
+  });
+
+  // two random points that are not equal, so are a valid input to twisted curve addition
+  let unequalPair = onlyIf(array(point, 2), ([p, q]) => !Curve.equal(p, q));
+
+  // test twisted curve gadgets witness generation
+
+  equivalentProvable({ from: [point], to: unit, verbose: true })(
+    (p) => Curve.isOnCurve(p) || throwError('expect on curve'),
+    (p) => CurveTwisted.assertOnCurve(p, Curve),
+    `${Curve.name} on curve`
+  );
+
+  equivalentProvable({ from: [unequalPair], to: point, verbose: true })(
+    ([p, q]) => Curve.add(p, q),
+    ([p, q]) => CurveTwisted.add(p, q, Curve),
+    `${Curve.name} add`
+  );
+
+  equivalentProvable({ from: [point], to: point, verbose: true })(
+    (p) => Curve.double(p),
+    (p) => CurveTwisted.double(p, Curve),
+    `${Curve.name} double`
+  );
+
+  equivalentProvable({ from: [point], to: point, verbose: true })(
+    Curve.negate,
+    (p) => CurveTwisted.negate(p, Curve),
+    `${Curve.name} negate`
+  );
+
+  equivalentProvable({ from: [point, scalar], to: point, verbose: true })(
+    (p, s) => {
+      let sp = Curve.scale(p, s);
+      return sp;
+    },
+    (p, s) => CurveTwisted.scale(s, p, Curve),
+    `${Curve.name} scale`
+  );
+
+  // test adding same point equals doubling
+  equivalentProvable({ from: [point], to: point, verbose: true })(
+    (p) => Curve.add(p, p),
+    (p) => CurveTwisted.double(p, Curve),
+    `${Curve.name} adding same point equals doubling`
+  );
+
+  equivalentProvable({ from: [point], to: point, verbose: true })(
+    (p) => Curve.double(p),
+    (p) => CurveTwisted.add(p, p, Curve),
+    `${Curve.name} doubling equals adding same point`
```

Comment:
I would add the following properties:
- scaling by the order gives the identity
- adding the identity gives the same point
- commutativity
- `r * (p + q) = r * p + r * q`
- `p + (q + s) = (p + q) + s`
- `(-(-p)) = p`
- `- identity` = `identity`
- test fromHex. In particular in the case that we don't have a point in the prime subgroup if we aim to be generic.

Reply:
will leave this 
> test fromHex. In particular in the case that we don't have a point in the prime subgroup if we aim to be generic.
for follow up as it is not clear that it won't be removed